src/lib.rs

```diff
@@ -0,0 +1,389 @@
+    pub fn new(ptn: &'a str) -> Result<Self, Box<dyn Error>> {
+        let options = FMOptions {
+            name_matcher: None,
+            ignore_leading_whitespace: true,
+            ignore_trailing_whitespace: true,
+        };
+        Ok(FMBuilder { ptn, options })
+    }
+
+    /// Add a name matcher `Some((ptn_re, text_re))` (or unset it with `None`). Defaults to `None`.
+    ///
+    /// Name matchers allow you to ensure that different parts of the text match without specifying
+    /// precisely what they match. For example, if you have output where you want to ensure that
+    /// two locations always match the same name, but the name is non-deterministic you can allow
+    /// the use of `$` wildcards in your pattern:
+    ///
+    /// ```rust
+    /// use {fm::FMBuilder, regex::Regex};
+    ///
+    /// let ptn_re = Regex::new("\\$.+?\\b").unwrap();
+    /// let text_re = Regex::new(".+?\\b").unwrap();
+    /// let matcher = FMBuilder::new("$1 b $1")
+    ///                         .unwrap()
+    ///                         .name_matcher(Some((ptn_re, text_re)))
+    ///                         .build()
+    ///                         .unwrap();
+    /// assert!(matcher.matches("a b a").is_ok());
+    /// assert!(matcher.matches("a b b").is_err());
+    /// ```
+    pub fn name_matcher(mut self, matcher: Option<(Regex, Regex)>) -> Self {
+        self.options.name_matcher = matcher;
+        self
+    }
+
+    #[doc(hidden)]
+    pub fn ignore_leading_whitespace(mut self, yes: bool) -> Self {
+        self.options.ignore_leading_whitespace = yes;
+        self
+    }
+
+    #[doc(hidden)]
+    pub fn ignore_trailing_whitespace(mut self, yes: bool) -> Self {
+        self.options.ignore_trailing_whitespace = yes;
+        self
+    }
+
+    /// Turn this `FMBuilder` into a `FMatcher`.
+    pub fn build(self) -> Result<FMatcher<'a>, Box<dyn Error>> {
+        Ok(FMatcher {
+            ptn: self.ptn,
+            options: self.options,
+        })
+    }
+}
+
+/// The fuzzy matcher.
+pub struct FMatcher<'a> {
+    ptn: &'a str,
+    options: FMOptions,
+}
+
+impl<'a> FMatcher<'a> {
+    /// A convenience method that automatically builds a pattern for you using `FMBuilder`'s
+    /// default options.
+    pub fn new(ptn: &'a str) -> Result<FMatcher, Box<dyn Error>> {
+        FMBuilder::new(ptn)?.build()
+    }
+
+    /// Does this fuzzy matcher match `text`? Returns `Ok(())` on success or `Err(usize)` on error
+    /// where the `usize` is the line number of the first line in `text` where the match failed.
+    pub fn matches(&self, text: &str) -> Result<(), usize> {
+        let mut names = HashMap::new();
+        let mut ptn_lines = self.ptn.lines();
+        let mut ptnl = ptn_lines.next();
+        let mut text_lines = text.lines();
+        let mut textl = text_lines.next();
+        let mut text_lines_off = 1;
+        loop {
+            match (ptnl, textl) {
+                (Some(x), Some(y)) => {
```

Comment:
I think we could improve some variable names for easier comprehension.

How about rename `ptnl` and `textl` to `ptn_it` and `text_it` and then `x` and `y` to `ptn_line` and `text_line`?

Reply:
(`it` as in iterator)

Reply:
I knew you would ask that here ;) Whatever naming scheme you choose is going to suck here. For example `ptn_it` is inaccurate (`ptnl` is a single line, not an iterator). I think we'll just get into bikeshedding if we're not careful.

Reply:
I misread again. It's not an iterator at all! Sorry.

Can we at least find better names than `x` and `y` or we may have to learn Fortran :P

Reply:
Unfortunately, I don't think we're going to agree on variable names: you like 'em long, I like 'em short. Given that the code in question comfortably fits on one screen, I think the short names are OK.

Reply:
How do you feel about shadowing `ptnl` and `textl`?

Reply:
That's worse IMHO.

Reply:
OK, I'll settle to agree to disagree on this one 😆